From 06fa21989ef6c9ff5623be019f56d2191a4c7bcb Mon Sep 17 00:00:00 2001
From: Zheng Yang <zhengyang@rock-chips.com>
Date: Wed, 15 Nov 2017 11:05:03 +0800
Subject: [PATCH] drm/rockchip: dw-hdmi: fix color depth check in YCBCR420 mode

If sink does't support YCBCR420 deep color, we return default
8bit.

In YCBCR420 mode, tmdsclock is half of RGB444 mode.

Change-Id: Ie3a1f8ca4bbe4b3bae5d7c9ea823fc798721a73a
Signed-off-by: Zheng Yang <zhengyang@rock-chips.com>
---
 drivers/gpu/drm/rockchip/dw_hdmi-rockchip.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/drivers/gpu/drm/rockchip/dw_hdmi-rockchip.c b/drivers/gpu/drm/rockchip/dw_hdmi-rockchip.c
index 63f8464073ec..6827ea7185c1 100644
--- a/drivers/gpu/drm/rockchip/dw_hdmi-rockchip.c
+++ b/drivers/gpu/drm/rockchip/dw_hdmi-rockchip.c
@@ -582,8 +582,12 @@ dw_hdmi_rockchip_check_depth(struct drm_display_info *info,
 		colordepth = depth;
 
 	/* Color depth on rockchip platform is limited up to 10bit */
-	if (colordepth > 10)
+	if (colordepth > 10) {
 		colordepth = 10;
+		if (color_format == DRM_HDMI_OUTPUT_YCBCR420 &&
+		    !(info->hdmi.y420_dc_modes & DRM_EDID_YCBCR420_DC_30))
+			return 8;
+	}
 
 	if (mode->flags & DRM_MODE_FLAG_DBLCLK)
 		pixclock *= 2;
@@ -596,6 +600,8 @@ dw_hdmi_rockchip_check_depth(struct drm_display_info *info,
 	else
 		tmdsclock = pixclock * colordepth / 8;
 
+	if (color_format == DRM_HDMI_OUTPUT_YCBCR420)
+		tmdsclock /= 2;
 	/*
 	 * For some display device, max_tmds_clock is 0, we think
 	 * max_tmds_clock is 340MHz. If tmdsclock > max_tmds_clock,
-- 
2.35.3

